File: 1710.maximum_units_on_a_truck.java
class Solution {
    public int maximumUnits(int[][] boxTypes, int truckSize) {
        int unitCount = 0;
        int remainingTruckSize = truckSize;
        while (remainingTruckSize > 0) {
            int maxUnitBoxIndex = findMaxUnitBox(boxTypes);
            // check if all boxes are used
            if (maxUnitBoxIndex == -1)
                break;
            // find the box count that can be put in truck
            int boxCount = Math.min(remainingTruckSize, boxTypes[maxUnitBoxIndex][0]);
            unitCount += boxCount * boxTypes[maxUnitBoxIndex][1];
            remainingTruckSize -= boxCount;
            // mark box with index maxUnitBoxIndex as used
            boxTypes[maxUnitBoxIndex][1] = -1;
        }
        return unitCount;
    }

    public int findMaxUnitBox(int[][] boxTypes) {
        int maxUnitBoxIndex = -1;
        int maxUnits = 0;
        for (int i = 0; i < boxTypes.length; i++) {
            if (boxTypes[i][1] > maxUnits) {
                maxUnits = boxTypes[i][1];
                maxUnitBoxIndex = i;
            }
        }
        return maxUnitBoxIndex;
}

class Solution {
    public int maximumUnits(int[][] boxTypes, int truckSize) {
        Arrays.sort(boxTypes, (a, b) -> b[1] - a[1]);
        int unitCount = 0;
        for (int[] boxType : boxTypes) {
            int boxCount = Math.min(truckSize, boxType[0]);
            unitCount += boxCount * boxType[1];
            truckSize -= boxCount;
            if (truckSize == 0)
                break;
        }
        return unitCount;
    }
}

File: 1710.maximum_units_on_a_truck.rb
# 1710.maximum_units_on_a_truck
# @param {Integer[][]} box_types
# @param {Integer} truck_size
# @return {Integer}
#
# brute force approach
# time: O(n^2) In the worst case we'd have to loop through every box 
#       type to get the max unit box index and then all remaining 
#       boxes if remaining matches the truck size. 
#
# space: O(1) We don't allocated any additional space and use the 
#       existing to mark what we have seen. 
# 
# note: This will get a TLE
def maximum_units(box_types, truck_size)
    unit_count = 0
    remaining_truck_size = truck_size

    while remaining_truck_size > 0 do
        max_unit_box_index = get_max_unit_box(box_types)

        break if max_unit_box_index == -1

        # clever trick to find to grab all that is remaining based on a constraint 
        box_count = [remaining_truck_size, box_types[max_unit_box_index][0]].min

        unit_count += box_count * box_types[max_unit_box_index][1]

        remaining_truck_size -= box_count

        box_types[max_unit_box_index][1] = -1

    end

    return unit_count
end

def get_max_unit_box(box_types)
    max_unit_box_index = -1
    max_units = 0

    box_types.each_with_index do |box, i|
        if box[1] > max_units
            max_units = box[1]
            max_unit_box_index = i
        end
    end

    return max_unit_box_index
end

# sort then iterate 
# time: O(n log n) we have to sort O(n log n) then we have to iterate O(n) thus,
#       O(n log n) + O(n) = O(n log n + n) = O(n log n)
#
# space: O(1) no additional structures added.  
def maximum_units(box_types, truck_size)
    unit_count = 0
    
    box_types.sort_by!{ |box| box[1] * -1} # we multiply by -1 to reverse the list
    
    box_types.each do |box|
        box_count = [truck_size, box[0]].min
        unit_count += box_count * box[1]
        truck_size -= box_count
        break if truck_size == 0
    end
   
    return unit_count
end


# max heap  
# time: O(n log n)
#
# space:O(n)
#
# note: not an optimization of the second approach and ruby does not contain a max heap i.e. priority queue
def maximum_units(box_types, truck_size)
    # TODO have to implement you're own PQ
end